src/bp/core/services/dialog/flow/service.ts

```diff
@@ -152,36 +126,66 @@ export class ScopedFlowService {
     private logger: Logger,
     private moduleLoader: ModuleLoader,
     private realtime: RealtimeService,
-    private botService: BotService
-  ) {}
+    private botService: BotService,
+    private invalidateFlow: (key: string, flow?: FlowView, newKey?: string) => void
+  ) {
+    this.cache = new ArrayCache<string, FlowView>(
+      x => x.name,
+      (x, prevKey, newKey) => ({ ...x, name: newKey, location: newKey })
+    )
+    this.expectedSavesCache = new LRUCache({ max: 100, maxAge: ms('20s') })
+  }
 
-  async handleInvalidatedCache(flowName: string) {
-    if (this._flowCache.isEmpty()) {
+  public localInvalidateFlow(key: string, flow?: FlowView, newKey?: string) {
+    if (!this.cache.values().length) {
       return
     }
 
+    if (flow) {
+      this.cache.update(key, flow)
+    } else if (newKey) {
+      this.cache.rename(key, newKey)
+    } else {
+      this.cache.remove(key)
+    }
+  }
+
+  public async handleInvalidatedCache(flowName: string) {
     const flowPath = this.toFlowPath(flowName)
+    const expectedSaves = this.expectedSavesCache.get(flowPath)
 
-    if (await this.ghost.fileExists(FLOW_DIR, flowPath)) {
-      const flow = await this.parseFlow(flowPath)
+    if (!expectedSaves) {
+      if (await this.ghost.fileExists(FLOW_DIR, flowPath)) {
+        const flow = await this.parseFlow(flowPath)
+        this.invalidateFlow(flowPath, flow)
+      } else {
+        this.invalidateFlow(flowPath, undefined)
+      }
 
-      this._flowCache.upsertFlow(this.botId, flow)
+      // TODO : this is kind of weird?
+      /*
+      // parent flows are only used by the NDU
+      if (this._isOneFlow()) {
+        const flows = this.cache.values()
+        const flowsWithParents = this.addParentsToFlows(flows)
+
+        this.cache.initialize(flowsWithParents)
+      }
+      */
```

Comment:
I'm not sure what this code is meant to do. I thought we dropped the ndu with albert, so what meaning does a parent flow have here?

Reply:
NDU is still supported and used by some clients/users. Since the computation of the parent flows was in O(n^2), I implemented a tree search wich is in O(logn). With this and the condition that validate if the current bot is a NDU one, there is practically no cost in keeping this block of code (+- 7ms for a bot with 3000+ flows).